### PR TITLE
feat: add support for publishing workers with r2 bucket bindings

### LINKED ISSUE #365: `r2` bucket support
Adding support for R2 bindings and commands, as defined in https://github.com/cloudflare/wrangler/commit/6408b15a4961d0bfa3934d76a540e2d1929f8c6d
- [x] https://github.com/cloudflare/wrangler2/pull/423
- [ ] https://github.com/cloudflare/wrangler2/pull/452

The implementation is the minimum to match what is provided by Wrangler 1 right now.
Before GA we should consider some polish such as:
- [ ] Ask before deleting a bucket
- [ ] Support bindings to preview buckets
- [ ] After bucket creation, provide guidance on how to configure a binding in the wrangler.toml

### PR DESCRIPTION
**Pull request by petebacondarwin**
This change adds the ability to define bindings in your `wrangler.toml` file
for R2 buckets. These buckets will then be available in the environment
passed to the worker at runtime.

Closes #365

### CODE REVIEW

File: packages/wrangler/src/config.ts [hunk 1]
```diff
@@ -203,6 +203,15 @@ export type Config = {
     preview_id?: string;
   }[];
 
+  r2_buckets?: {
+    /** The binding name used to refer to the R2 bucket in the worker. */
+    binding: string;
```
Comment (threepointone) [inline packages/wrangler/src/config.ts:208]: really dislike "binding" bah. I'll make some noise about it later.

**Review by threepointone — CHANGES_REQUESTED** [review]
This looks good to me! Could you add it to `wrangler dev` as well? That would be needed to close the issue.

**Comment by threepointone** [comment]
You could also add `wrangler dev` support in another PR, but then remove the "Closes" directive from the PR description, and I'll stamp this.

**Comment by petebacondarwin** [comment]
That is so weird. I actually ran `wrangler dev` on my test project and I saw the R2 binding in the worker....
But looking through the code, again, I can see that it is not being added at the `<Dev>` component...
Is there any reason why dev extracts these bindings differently to how publish? Perhaps there is an opportunity for more de-duplication there?

**Comment by petebacondarwin** [comment]
Looking through the code where `toFormData()` is used, it seems that we also need to update when submitting a secret. (Not quite sure why... is that related to the issue before of setting secrets where wiping out other environment variables?)

**Comment by petebacondarwin** [comment]
Added the bindings to `wrangler dev` I'll check it manually later today.
I also changed the `CfWorkerInit` so that each of the `bindings` properties were required (but could still be undefined) so as to ensure we don't miss this again.
@threepointone - PTAL.

**Review by threepointone — APPROVED** [review]
image

**Comment by threepointone** [comment]
> Is there any reason why dev extracts these bindings differently to how publish? Perhaps there is an opportunity for more de-duplication there?

Probably! It'll also force people to ship support for both whenever they make a new type of binding.

**Pull request merged**